Avoid packaging / unpacking cycle when using local current distributions

breskeby opened the pull request:
When test cluster or local distribution setups (e.g. `:run` or `:localDistro`) rely on local build distributions we have the overhead of packing and unpacking the distribution archive tars or zips. This PR removes this overhead by registering a directory backed variant of a distribution per distribution type.

We introduce a simple DSL for declaring distribution_archives that is used as base for the archive tasks creating zips and tars _AND_ for the `explodedDist` tasks that only do a simple copy.

Along the way common configurations in `distribution/archives` have been moved into an internal plugin to
- improve test coverage for common configurations
- Removes build script cluttering

Here are Comparing timelines before and after this change:

* before: https://gradle-enterprise.elastic.co/s/3b64ptmp4zugg/timeline 
* after: https://gradle-enterprise.elastic.co/s/ezwhjyrl3dnfw/timeline

Benchmarking this change on OSX (relying on `./gradlew :localDistro` with a non ABI change in `:server`) we get an average improvement of 18.5s: 

<img width="630" alt="Screenshot 2020-09-02 at 18 28 34" src="https://user-images.githubusercontent.com/77300/92011028-c8c5f200-ed4a-11ea-8a92-1f4f04184924.png">

breskeby committed:
Extract distribution archives defaults into plugin

Extracting more stuff from the distribution/archives scripts into a testable plugin

- Added basic test coverage

breskeby committed:
Setup local distributions faster

- avoid packaging/unpackaging cycle when relying on locally build distributions
- Move more logic into distribution/archives

breskeby committed:
Provide DSL for setting up distribution archives

- provide directory variant per distribution archive

breskeby committed:
Cleanup archives build script

breskeby committed:
Fix consuming dependencies to distribution archives

- polishing

breskeby committed:
Do not resolve dependencies in toString

breskeby committed:
Do not rely on toString to resolve file path

breskeby committed:
Explicitly set dist config artifact format

breskeby committed:
Only declare explicit artifact format for extractable distributions

breskeby committed:
Do not rely on toString in DistroPlugin

breskeby commented on distribution/docker/build.gradle:
```diff
@@ -146,8 +155,6 @@ void addCopyDockerContextTask(Architecture architecture, boolean oss, DockerBase
       }
     }
-
-    from configurations.dockerPlugins
```
didn't find any use of this one

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Build)

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/DistributionArchive.java:
```diff
@@ -0,0 +1,65 @@
+        return explodedDistTask;
+    }
+
+    interface ContentProvider {
```
Do we really need this interface? Could we not just use `Supplier<CopySpec>`?

breskeby replied:
fixed

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionArchiveSetupPlugin.java:
```diff
@@ -0,0 +1,196 @@
+ * Elasticsearch distribution archives. See ':distribution:archives'.
+ *
+ * This configures the default artifacts for the distribution specific
+ * subprojects. We have subprojects for two reasons:
```
4. Having per-distribution sub-projects means we can build them in parallel.

breskeby replied:
Added

mark-vieira commented on distribution/archives/build.gradle:
```diff
@@ -17,39 +17,11 @@
-}
+apply plugin: 'elasticsearch.internal-distribution-archive-setup'
 
 CopySpec archiveFiles(CopySpec modulesFiles, String distributionType, String platform, String architecture, boolean oss, boolean jdk) {
```
I would love for us to either a) make a proper DSL from this or b) use named parameters when calling this method rather than have `foo(true, false, null, false)` all over the place which is a poor description of what is happening.

breskeby replied:
a proper DSL is currently a bit tricky as bits of that copySpec even live in the distribution package one level above. I see what I can do to make it a bit more readable though (named parameters sound like a good idea for now)

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionArchiveSetupPlugin.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    private String buildTaskToSubprojectName(String taskName) {
+        return taskName.substring("build".length()).replaceAll("[A-Z]", "-$0").toLowerCase().substring(1);
```
I'm not sure we need the "build" prefix convention here for container entries. The DSL is meant to describe a distribution, the the task that eventually constructs it. The latter is an implementation detail handled by the plugin. I think it makes more sense to create a `darwinTar` and then the plugin incidentally creates a `buildDarwinTar` task.

breskeby replied:
fixed

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionArchiveSetupPlugin.java:
```diff
@@ -0,0 +1,196 @@
+                var publications = defaultConfiguration.getOutgoing();
+                var variant = (ConfigurationVariantInternal) publications.getVariants().maybeCreate("directory");
+                variant.getAttributes().attribute(ArtifactAttributes.ARTIFACT_FORMAT, ArtifactTypeDefinition.DIRECTORY_TYPE);
+                variant.artifactsProvider(() -> Collections.singletonList(new DirectoryPublishArtifact(explodedArchiveTask)));
```
Is there no public API to accomplish this?

breskeby replied:
Unfortunately there isn't as far as I can tell. I also reached out to the gradle folks in slack and in the forum at https://discuss.gradle.org/t/registering-a-directory-type-variant-backed-by-a-custom-copy-task/37486/2 but so far no feedback

breskeby replied:
I changed the implementation here to rely on a different configuration for the explodedDist. This allows us to not rely on internal api and makes the setup a bit simpler

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/internal/InternalDistributionArchiveSetupPlugin.java:
```diff
@@ -0,0 +1,196 @@
+        // create the directory that we want, and then point CopySpec to its
+        // parent to copy to the root of the distribution
+        File logsDir = new File(project.getBuildDir(), "logs-hack/logs");
+        project.getExtensions().add("logsDir", new File(project.getBuildDir(), "logs-hack/logs"));
```
Shouldn't these be extra properties instead of extensions? I assume there's not too much difference but extensions end up get decorated by Gradle at runtime which we probably don't need in this case.

breskeby replied:
good point. fixed.

mark-vieira reviewed (approved):
I left a few comments but otherwise this looks great. This is so much more efficient for local developoment. Well done!

breskeby committed:
Apply review feedback

- expose explodedDist in separate configuration
- redo namings
- minor renamings

breskeby committed:
Fix docker sources

breskeby committed:
Fix imports

breskeby committed:
Explicitly request default configuratino for docker sources

breskeby commented on buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:
```diff
@@ -124,13 +124,22 @@ void setupDistributions(Project project) {
         }
     }
 
+    private String resolveArtifactFormat(ElasticsearchDistribution distribution) {
```
changed

breskeby commented on buildSrc/src/main/java/org/elasticsearch/gradle/DistributionDownloadPlugin.java:
```diff
@@ -124,13 +124,22 @@ void setupDistributions(Project project) {
             dependencies.add(distribution.configuration.getName(), resolvedDependency);
             // no extraction allowed for rpm, deb or docker
             if (distribution.getType().shouldExtract()) {
+                distribution.configuration.getAttributes()
```
I changed the providing setup to rely on different configurations instead (default and extended). This means we do not need that bit anymore

breskeby commented:
@elasticmachine test this please

breskeby commented:
@elasticmachine test this please

breskeby commented:
@elasticmachine test this please

breskeby committed:
Fix resolveAllDependencies

when relying on variants and artifacts we must make the requested configuration explicit

breskeby committed:
Minor polishing
